### ISSUE TITLE
segfault on ubuntu

### ISSUE THREAD

**Issue by flamholz** [debbiemarkslab/plmc#1]
regardless of how I compile, pimc was segfaulting for me on ubuntu today running the DHFR example. will provide details on ubuntu and GCC versions from lab tomorrow.

avi

**Comment by jingraham** [comment]
hi avi,

thanks for raising this. we recently heard of a similar issue from someone running ubuntu and were unsuccessful at reproducing on an ubuntu machine in lab. it would be great to get the hardware specs (amount of RAM + CPU architecture) along with the ubuntu and gcc versions. some current guesses are either memory issues or compilation options (-O3 -msse4.2 flags might not be portable), though maybe only valgrind will tell

john

**Comment by flamholz** [comment]
Still segfaults if I make clean and then make all without the -O3 -msse4.2 flags (commented out in the GCCFLAGS in the makefile). I will try to run a system update and see if these errors dissappear - i.e. if they are related to the GCC version. 

Computer Details:

Ubuntu 14.04 LTS 64-bit
gcc 4.84
17.6 GB RAM
Intel Core i7 CPU 920, 8 x 2.67 GHx

**Comment by jingraham** [comment]
Ah, sounds like it will be trickier then. Thanks for checking that. Do you have the output from plmc before the segfault?

**Comment by flamholz** [comment]
Apparently I'm on the latest gcc version for Ubuntu 14.04 (at least the latest one that apt-get will install for me). I just installed clang and rebuilt using "make all-mac" the binary built by clang also segfaults, which leads me to believe that it is related to how your code uses some system library that differs between os x and ubuntu. 

I will post the output but I need to log in from the linux box for that.
A

**Comment by jingraham** [comment]
Good idea to test that. The confusing bit is that we tend to run this code on Linux in our lab, mainly on CentOS and sometimes on Ubuntu. I will try to see if I can reproduce on a virtual machine with that GCC + Ubuntu combination. Thanks for your patience with the bugs - John

**Comment by flamholz** [comment]
Compiled with debug flags and ran inside gdb to produce a stack trace. Here's the output 


> savage@savage-Studio-XPS-435T:~/workspace/plmc$ gdb --args bin/plmc -o example/protein/DHFR.params -le 16.0 -lh 0.01 -m 100 -g -f DYR_ECOLI example/protein/DHFR.a2m
> GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.2) 7.7.1
> Copyright (C) 2014 Free Software Foundation, Inc.
> License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
> This is free software: you are free to change and redistribute it.
> There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
> and "show warranty" for details.
> This GDB was configured as "x86_64-linux-gnu".
> Type "show configuration" for configuration details.
> For bug reporting instructions, please see:
> <http://www.gnu.org/software/gdb/bugs/>.
> Find the GDB manual and other documentation resources online at:
> <http://www.gnu.org/software/gdb/documentation/>.
> For help, type "help".
> Type "apropos word" to search for commands related to "word"...
> Reading symbols from bin/plmc...done.
> (gdb) run
> Starting program: /home/savage/workspace/plmc/bin/plmc -o example/protein/DHFR.params -le 16.0 -lh 0.01 -m 100 -g -f DYR_ECOLI example/protein/DHFR.a2m
> Found focus DYR_ECOLI as sequence 1
> 3616 valid sequences out of 3629 
> 159 sites out of 171
> Region starts at 1
> Effective number of samples: 1568.1	(80% identical neighborhood = 1.000 samples)
> 
> Program received signal SIGSEGV, Segmentation fault.
> 0x0000000000409627 in InferPairModel (ali=0x6162b0, options=0x616010)
>     at src/inference.c:103
> 103	                xHi(i, seq(s, i)) += ali->weights[s] * Zinv;
> (gdb) bt
> #0  0x0000000000409627 in InferPairModel (ali=0x6162b0, options=0x616010)
>     at src/inference.c:103
> #1  0x0000000000404ea7 in main (argc=13, argv=0x7fffffffde88) at src/plm.c:196

**Comment by jingraham** [comment]
Thanks for stack trace! I think that got it. I reproduced the issue on a VM + valgrind and it seems the parameter initialization in 103 was initializing gap parameters that no longer exist in gap-ignoring mode (-g). Fortunately, since our problem is convex, the initialization only affects speed of convergence and not accuracy. I patched it [here](https://github.com/debbiemarkslab/plmc/commit/18a5d8d60d795d17889dd54cc2d496608dda99cc), which fixed the issue on Ubuntu 14 + gcc 4.4 for me. How does that work for you?

john

**Comment by flamholz** [comment]
Seems to get rid of the segfault on the DHFR test on my linux box! Running
multicore on this machine's 8 cores and 18 GB RAM is nice. Feel free to
close this bug.
A

On Fri, Dec 23, 2016 at 8:05 AM John Ingraham <notifications@github.com>
wrote:

Thanks for stack trace! I think that got it. I reproduced the issue on a VM
+ valgrind and it seems the parameter initialization in 103 was
initializing gap parameters that no longer exist in gap-ignoring mode (-g).
Fortunately, since our problem is convex, the initialization only affects
speed of convergence and not accuracy. I patched it here
<https://github.com/debbiemarkslab/plmc/commit/18a5d8d60d795d17889dd54cc2d496608dda99cc>,
which fixed the issue on Ubuntu 14 + gcc 4.4 for me. How does that work for
you?

john

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/debbiemarkslab/plmc/issues/1#issuecomment-269011030>,
or mute the thread
